core/src/Kokkos_MpiShmem.hpp

```diff
@@ -0,0 +1,223 @@
+//@HEADER
+*/
+
+#ifndef KOKKOS_MPISHMEM_HPP
+#define KOKKOS_MPISHMEM_HPP
+
+#include <Kokkos_Core_fwd.hpp>
+
+// If MPI execution space is enabled then use this header file.
+
+#if defined( KOKKOS_HAVE_MPISHMEM )
+
+#include <iosfwd>
+
+#include <Kokkos_MpiShmemSpace.hpp>
+
+#include <Kokkos_Parallel.hpp>
+#include <Kokkos_Layout.hpp>
+#include <Kokkos_ScratchSpace.hpp>
+#include <Kokkos_MemoryTraits.hpp>
+#include <impl/Kokkos_Tags.hpp>
+
+/*--------------------------------------------------------------------------*/
+
+namespace Kokkos {
+namespace Experimental {
+namespace Impl {
+class MpiShmemExec ;
+} // namespace Impl
+} // namespace Experimental
+} // namespace Kokkos
+
+/*--------------------------------------------------------------------------*/
+
+namespace Kokkos {
+namespace Experimental {
+
+/// \class MpiShmem
+/// \brief Kokkos Execution Space that uses MPI shared memory communicators.
+///
+/// An "execution space" represents a parallel execution model.  It tells Kokkos
+/// how to parallelize the execution of kernels in a parallel_for or
+/// parallel_reduce.  For example, the Threads execution space uses Pthreads or
+/// C++11 threads on a CPU, the OpenMP execution space uses the OpenMP language
+/// extensions, and the Serial execution space executes "parallel" kernels
+/// sequentially.  The MpiShmem execution space uses MPI's
+/// shared memory communicator splitting to execute kernels across MPI
+/// ranks on the same CPU.
+class MpiShmem {
+public:
+  //! \name Type declarations that all Kokkos execution spaces must provide.
+  //@{
+
+  //! Tag this class as a kokkos execution space
+  typedef MpiShmem              execution_space ;
+
+  //! This execution space's preferred memory space.
+  typedef MpiShmemSpace         memory_space ;
+
+  //! This execution space preferred device_type
+  typedef Kokkos::Device<execution_space,memory_space> device_type;
+
+  //! The size_type best suited for this execution space.
+  typedef memory_space::size_type  size_type ;
+
+  //! This execution space's preferred array layout.
+  typedef LayoutRight            array_layout ;
+
+  //!
+  typedef ScratchMemorySpace< MpiShmem >  scratch_memory_space ;
+
+  //@}
+  //--------------------------------------------------
+  //! \name Functions that all Kokkos devices must implement.
+  //@{
+
+  /// \brief True if and only if this method is being called in a
+  ///   thread-parallel function.
+  inline static int in_parallel() {
+    TODO!!!
+  }
+
+  /** \brief  Set the device in a "sleep" state. A noop for MpiShmem.  */
+  static bool sleep();
+
+  /** \brief Wake the device from the 'sleep' state. A noop for MpiShmem. */
+  static bool wake();
+
+  /// \brief Wait until all dispatched functors complete.
+  ///
+  /// This is currently a noop for MpiShmem, although we could
+  /// do something like MPI_Ireduce in the future.
+  static void fence();
+
+  //! Free any resources being consumed by the device.
+  static void finalize();
```

Comment:
Please don't call `MPI_Finalize` here.  Mike Heroux prefers us to think of MPI as owned by the application, and ourselves as a library that uses MPI.  Thus, the application is responsible for calling `MPI_Init` and `MPI_Finalize`; Kokkos is not.